Raise appropriate error message when append is not possible

j-c-cook opened the pull request:
- Handle "append" by adding **options to SqlWriter 
- Check logger instance for SqliteWriter in rotating logger setup and raise appropriate error.

closes #1318

j-c-cook commented:
@zariiii9003 ready for review.

zariiii9003 commented on can/io/sqlite.py:
```diff
@@ -128,7 +128,9 @@ class SqliteWriter(MessageWriter, BufferedReader):
 
-    def __init__(self, file: StringPathLike, table_name: str = "messages") -> None:
+    def __init__(
+        self, file: StringPathLike, table_name: str = "messages", **options: Any
```
```suggestion
        self, file: StringPathLike, table_name: str = "messages", **kwargs: Any
```

This is usually called `kwargs` in python projects

j-c-cook replied:
> Also: Shouldn't there be an exception when the user demands "append" but the logger does not support it?

I think this is brilliant.

zariiii9003 commented on can/io/logger.py:
```diff
@@ -225,6 +226,11 @@ def _get_new_writer(self, filename: StringPathLike) -> FileIOMessageWriter:
+            )
         else:
             raise Exception(
                 "The Logger corresponding to the arguments is not a FileIOMessageWriter or "
```
You could improve the existing error message without the `isinstance(logger, can.io.SqliteWriter)` check. You can access the logger class name with `logger.__class__.__name__`, for example inside a f-string.
This way you do not need special cases for any new unsupported loggers

j-c-cook commented:
@zariiii9003 Thank you for that feedback. I added in the "append" error handling to `ASCWriter` and `SQLWriter` and adjusted the exception based on your [comment](https://github.com/hardbyte/python-can/pull/1361#discussion_r939557531).

j-c-cook commented on can/io/asc.py:
```diff
@@ -360,6 +361,13 @@ def __init__(
         """
+        if "append" in kwargs:
+            if kwargs["append"]:
+                raise TypeError(
```
I'm not sure if this is the best kind of error to use. Maybe a `ValueError`?

zariiii9003 replied:
Yes, I'd say ValueError, too

zariiii9003 commented on can/io/logger.py:
```diff
@@ -227,8 +228,8 @@ def _get_new_writer(self, filename: StringPathLike) -> FileIOMessageWriter:
             raise Exception(
-                "The Logger corresponding to the arguments is not a FileIOMessageWriter or "
-                "can.Printer"
+                f"The Logger, {logger.__class__.__name__}, corresponding to "
```
The previous error message wasn't great. We should use simple language.

```python3
raise Exception(
    f"The log format \"{''.join(pathlib.Path(filename).suffixes[-2:])}\" "
    f"is not supported by {self.__class__.__name__}"
)
```
This results in `Exception: The log format ".db" is not supported by SizedRotatingLogger`.
Works for all subclasses, too. We could also point the user to the documentation but that might be a bit redundant.

zariiii9003 commented on can/io/asc.py:
```diff
@@ -360,6 +361,13 @@ def __init__(
                         have a channel set
         """
+        if "append" in kwargs:
+            if kwargs["append"]:
```
you could remove this line if you use `if kwargs.get("append", False):`

j-c-cook commented on can/io/asc.py:
```diff
@@ -361,13 +358,11 @@ def __init__(
-                    f"equipped to append messages to an existing "
-                    f"file."
-                )
+        if kwargs.get("append", False):
+            raise TypeError(
+                f"{self.__class__.__name__} is currently not equipped to "
+                f"append messages to an existing file."
+            )
```
Took your advice here.

j-c-cook commented on can/io/logger.py:
```diff
@@ -228,7 +228,7 @@ def _get_new_writer(self, filename: StringPathLike) -> FileIOMessageWriter:
         else:
             raise Exception(
-                f"The log format \".{filename.split('.')[-1]}\" is not "
+                f"The log format \"{os.path.splitext(filename)[-1]}\" is not "
```
I used your recommended error message, but slightly adjusted how the extension is found.

zariiii9003 replied:
This does not work for multiple suffixes like `file.asc.gz`

j-c-cook replied:
I fixed it with your version.

j-c-cook commented:
@zariiii9003 I think this might be ready to go in. Could you give it another look?

zariiii9003 reviewed (approved):
Thank you, looks good
